### ISSUE TITLE
Keyboard shortcuts, media keys in particular

### ISSUE THREAD

**Issue by aidenlangley** [Spotifyd/spotifyd#1015]
**Is your feature request related to a problem? Please describe.**
Would like to be able to use media keys to resume/pause playback, forward/back track, etc.

**Describe the solution you'd like**
Keyboard shortcut support.

**Describe alternatives you've considered**
Other Spotify clients. :)

**Comment by YodaEmbedding** [comment]
The way to do this is supporting dbus volume controls.

```
λ playerctl -p spotifyd volume 0.05-
Could not execute command: GDBus.Error:org.freedesktop.DBus.Error.Failed: Property Volume is not writable
```

**Comment by eladyn** [comment]
If someone wants to implement this, [here](https://github.com/Spotifyd/spotifyd/blob/603bcc4d802d2b8f45ac517c0513eb4096252d7b/src/dbus_mpris.rs#L297-L306) is the relevant code which would need to call the `.set()` method as well. You can also have a look at how [ncspot does it](https://github.com/hrkfdn/ncspot/blob/f29d263f41e0424ae42071f684c82fffab3afcf5/src/mpris.rs#L320-L340).

**Comment by aidenlangley** [comment]
Would you be able to give me a brief explanation of this code @eladyn ?

```rs
let mv_api_token = api_token.clone();
b.property("Volume").emits_changed_false().get(move |_, _| {
    let sp = create_spotify_api(&mv_api_token);
    let vol = sp
        .current_playback(None)
        .ok()
        .flatten()
        .map_or(0.0, |p| p.device.volume_percent as f64);
    Ok(vol)
});
```

Notable differences in this code and the `ncspot` example appear to be `.access(Access::ReadWrite).on_get` in their example, and the absence of `.emits_changed_false()` in their example.

**Comment by eladyn** [comment]
**Note:** I just reread your original issue description and it seems that the missing volume control isn't really your problem. If you use Linux, media keys should already work in most desktop environments if you have compiled `spotifyd` with the `dbus_mpris` feature flag and enabled it in the config (`use_mpris = true`). If this is the case, I'm not sure, wether the following will help you. If you are still interested in Volume control in the standard MPRIS way, you can of course read it.

Sure, going to try my best.

The reason for the difference is apparently that `spotifyd` uses [`dbus_crossroads`](https://crates.io/crates/dbus-crossroads), whereas `ncspot` uses [`dbus_tree`](https://crates.io/crates/dbus-tree) for constructing the interface.

You can find more or less complete documentation for the `dbus_crossroads` crate [here](https://docs.rs/dbus-crossroads/0.5.0/dbus_crossroads/index.html), although you probably want to start with the [`IfaceBuilder`](https://docs.rs/dbus-crossroads/0.5.0/dbus_crossroads/struct.IfaceBuilder.html), since this is the type of the variable `b` in this scope, which is used to add all the methods and properties.

Calling the `.property()` method returns a builder for the Property ([`PropBuilder`](https://docs.rs/dbus-crossroads/0.5.0/dbus_crossroads/struct.PropBuilder.html)), which uses the typical Builder pattern (see e.g. [here](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html), if you haven't yet stumbled over this).

The different [`.emits_changed_*()`](https://docs.rs/dbus-crossroads/0.5.0/dbus_crossroads/struct.PropBuilder.html#method.emits_changed_false) methods are used to express, wether we plan to emit a `PropertiesChanged` event, when we change the property. Thus, in this case, you'd want to set it to `true`, since the MPRIS spec tells us, to [emit such signal](https://specifications.freedesktop.org/mpris-spec/latest/Player_Interface.html#Property:Volume).

The [`.get()`](https://docs.rs/dbus-crossroads/0.5.0/dbus_crossroads/struct.PropBuilder.html#method.get) and the [`.set()`](https://docs.rs/dbus-crossroads/0.5.0/dbus_crossroads/struct.PropBuilder.html#method.set) methods can be used, to provide the interface with callbacks that it can call, when someone is trying to get or set the property over dbus. In most cases, you will want to use a closure here instead of a "real" function.

Currently, the `.get()` method is implemented, but the `.set()` is missing, so this property is readonly. You'd need to append a [`.set(...)`](https://docs.rs/dbus-crossroads/0.5.0/dbus_crossroads/struct.PropBuilder.html#method.set) afterwards to handle a volume change.
You can pass this method a closure that takes three arguments. You can probably ignore the first two, since the third contains the volume, the user wants to change to. To actually change the volume, the [`.volume(...)`](https://docs.rs/rspotify/0.10.0/rspotify/client/struct.Spotify.html#method.volume) method from `rspotify` might interest you.

Hope, this helps you with your problem.

**Comment by aidenlangley** [comment]
Thank you for the detailed explanation, much appreciated.

You're right, the original issue isn't quite what I'd believed it was, but spotifyd's behaviour is not what you'd expect it to be - usually `playerctl pause` would pause media playback without the need to specify which media player you'd like to pause. Having a separate case, and keybinding, for spotifyd, is not how this ought to work in my opinion.

Not sure how the volume issue came up, seems mostly unrelated but I guess it is relevant as it also requires a not-so-standard approach.

**Comment by eladyn** [comment]
You're welcome!

I'm not sure what you mean. As long as `spotifyd` is the only player running, `playerctl pause/play/play-pause` works just fine for me. (If multiple players are active, `playerctl` may of course select a different one.) 

Are you able to control `spotifyd` by actively selecting it in `playerctl` (with the `-p spotifyd` flag) and do you have more than one player active (`playerctl --list-all`)?

**Comment by aidenlangley** [comment]
![2022-01-20_09-30](https://user-images.githubusercontent.com/684721/150208907-470f2894-6b89-4ca8-94be-36099716189e.png)

This is what I'm experiencing ^

I download the full release directly from the GitHub releases, and then use `spotify-tui` to control playback instead of `playerctl`.

**Comment by eladyn** [comment]
That is really interesting, since it works for me out of the box. Apparently `spotifyd` is correctly selected by `playerctl`, but it somehow fails to successfully execute the command.

Some ideas:

- You can try if [enabling debug logging](https://github.com/altdesktop/playerctl#debug-logging) for `playerctl` shows interesting messages.
- Does executing `dbus-send --session --print-reply --dest=org.mpris.MediaPlayer2.spotifyd /org/mpris/MediaPlayer2 org.mpris.MediaPlayer2.Player.Play` directly work?
- Do any of the other available commands in `playerctl` work or do they all fail with the same message? (`metadata`, `status`, ...)
- If nothing else helps, maybe there is something interesting in the `spotifyd --verbose` logs?

**Comment by aidenlangley** [comment]
```sh
~ ❱ G_MESSAGES_DEBUG=playerctl playerctl play
(playerctl:356781): playerctl-DEBUG: 10:29:47.713: Getting list of player names from D-Bus
(playerctl:356781): playerctl-DEBUG: 10:29:47.713: Getting list of player names from D-Bus
(playerctl:356781): playerctl-DEBUG: 10:30:12.716: initializing player: spotifyd
No player could handle this command
~ 25s [1] dbus-send --session --print-reply --dest=org.mpris.MediaPlayer2.spotifyd /org/mpris/MediaPlayer2 org.mpris.MediaPlayer2.Player.Play
Error org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
```

Maybe I am missing a piece of MPRIS? Works okay when Firefox is available to control.

```sh
~ ❱ G_MESSAGES_DEBUG=playerctl playerctl play
(playerctl:378692): playerctl-DEBUG: 10:34:00.910: Getting list of player names from D-Bus
(playerctl:378692): playerctl-DEBUG: 10:34:00.910: Getting list of player names from D-Bus
(playerctl:378692): playerctl-DEBUG: 10:34:00.911: initializing player: firefox.instance2483
```

**Comment by eladyn** [comment]
Hmm, that looks interesting. I assume that those `25s` at the beginning of the line report the execution time of the previous command? If so, then `playerctl` as well as the `dbus-send` call is apparently running into a timeout.

Even after looking at the code, I currently have no idea, why this might be happening. Maybe you could try adding some debug logging into `dbus_mpris.rs` and compile `spotifyd` from source, to investigate where / if it gets stuck somewhere?

I'd propose for example the following places: [dbus_server creation begins](https://github.com/Spotifyd/spotifyd/blob/603bcc4d802d2b8f45ac517c0513eb4096252d7b/src/dbus_mpris.rs#L114), [`.Play`-method callback](https://github.com/Spotifyd/spotifyd/blob/603bcc4d802d2b8f45ac517c0513eb4096252d7b/src/dbus_mpris.rs#L205), [creation of the server is finished](https://github.com/Spotifyd/spotifyd/blob/603bcc4d802d2b8f45ac517c0513eb4096252d7b/src/dbus_mpris.rs#L461)

**Comment by aidenlangley** [comment]
Yes 25s is how long it took, same amount of time on both occasions.

Sure, I'll debug, happy to work on this, thanks for your help so far.

**Comment by aidenlangley** [comment]
So I've discovered that it works as you've described when the binary is located in `/usr/bin` as opposed to `$HOME/.local/bin`. `spt` doesn't seem to mind as it interacts directly with an `rspotify::Spotify` client, so haven't noticed this before.

**Comment by eladyn** [comment]
Everything mentioned here should be fixed by now. Closing...
